### PR TITLE
Use the new networkstatus flow to know if we have internet or not

### PR DESCRIPTION
**Pull request by tevincent**
depends on https://github.com/Infomaniak/android-core/pull/208

https://github.com/user-attachments/assets/1eb2241f-ea3b-4972-b787-61d69e60bc36

### CODE REVIEW

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 1]
```diff
@@ -185,6 +189,22 @@ class MainViewModel @Inject constructor(
...
+                Breadcrumb().apply {
+                    category = "Network"
+                    message = "Internet access is available : $isAvailable"
+                    level = if (isAvailable) SentryLevel.INFO else SentryLevel.WARNING
+                }.also(Sentry::addBreadcrumb)
+
+                isInternetAvailable.postValue(isAvailable)
+            }
+        }
+    }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:206]:
```suggestion
    fun observeNetworkStatus() = viewModelScope.launch(Dispatchers.IO) {
        appContext.networkStatusFlow().collect { isAvailable ->
            SentryLog.d("Internet availability", if (isAvailable) "Available" else "Unavailable")

            Breadcrumb().apply {
                category = "Network"
                message = "Internet access is available : $isAvailable"
                level = if (isAvailable) SentryLevel.INFO else SentryLevel.WARNING
            }.also(Sentry::addBreadcrumb)

            isInternetAvailable.postValue(isAvailable)
        }
    }
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt [hunk 2]
```diff
@@ -76,8 +78,14 @@ class DownloadAttachmentProgressDialog : DialogFragment() {
     }
 
     private fun popBackStackWithError() {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:80]:
```suggestion
    private fun popBackStackWithError() = lifecycleScope.launch(Dispatchers.IO) {
```

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt [hunk 3]
```diff
@@ -76,8 +78,14 @@ class DownloadAttachmentProgressDialog : DialogFragment() {
...
-        findNavController().popBackStack()
+        lifecycleScope.launch(Dispatchers.IO) {
+            requireContext().networkStatusFlow().collect { isInternetAvailable ->
+                val title = if (isInternetAvailable) R.string.anErrorHasOccurred else R.string.noConnection
+                withContext(Dispatchers.Main) {
+                    showSnackbar(title)
+                    findNavController().popBackStack()
+                }
+            }
+        }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:89]:
```suggestion
        requireContext().networkStatusFlow().collect { isInternetAvailable ->
            val title = if (isInternetAvailable) R.string.anErrorHasOccurred else R.string.noConnection
            withContext(Dispatchers.Main) {
                showSnackbar(title)
                findNavController().popBackStack()
            }
        }
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Review by FabianDevel — DISMISSED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 4]
```diff
@@ -169,6 +169,17 @@ class MainViewModel @Inject constructor(
...
 
+    val isNetworkAvailable = NetworkAvailability(appContext).isNetworkAvailable
+        .mapLatest {
+            SentryLog.d("Internet availability", if (it) "Available" else "Unavailable")
+            it
+        }
+        .stateIn(
+            scope = viewModelScope,
+            started = SharingStarted.Eagerly,
+            initialValue = null
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:180]:
```suggestion
            initialValue = null,
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 5]
```diff
@@ -495,8 +497,7 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
             SwipeAction.NONE -> error("Cannot swipe on an action which is not set")
         }
 
-        val shouldKeepItemBecauseOfNoConnection = isInternetAvailable.value == false
-
+        val shouldKeepItemBecauseOfNoConnection = mainViewModel.isNetworkAvailable.value == false
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:500]:
```suggestion
        val shouldKeepItemBecauseOfNoConnection = isNetworkAvailable.value == false
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 6]
```diff
@@ -529,10 +530,14 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
     }
 
     private fun observeNetworkStatus() {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:532]:
```suggestion
    private fun observeNetworkStatus() = lifecycleScope.launch {
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 7]
```diff
@@ -529,10 +530,14 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
+        lifecycleScope.launch {
+            mainViewModel.isNetworkAvailable.collect { isNetworkAvailable ->
+                if (isNetworkAvailable != null) {
+                    TransitionManager.beginDelayedTransition(binding.root)
+                    binding.noNetwork.isGone = isNetworkAvailable
+                    if (!isNetworkAvailable) updateThreadsVisibility()
+                }
+            }
         }
     }
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:542]:
```suggestion
        mainViewModel.isNetworkAvailable.collect { isNetworkAvailable ->
            if (isNetworkAvailable != null) {
                TransitionManager.beginDelayedTransition(binding.root)
                binding.noNetwork.isGone = isNetworkAvailable
                if (!isNetworkAvailable) updateThreadsVisibility()
            }
        }
    }
```

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt [hunk 8]
```diff
@@ -529,10 +530,14 @@ class ThreadListFragment : TwoPaneFragment(), SwipeRefreshLayout.OnRefreshListen
...
 
     private fun observeNetworkStatus() {
-        mainViewModel.isInternetAvailable.observe(viewLifecycleOwner) { isAvailable ->
-            TransitionManager.beginDelayedTransition(binding.root)
-            binding.noNetwork.isGone = isAvailable
-            if (!isAvailable) updateThreadsVisibility()
+        lifecycleScope.launch {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:533]: Can you please specify the Dispatcher ?

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt [hunk 9]
```diff
@@ -76,8 +79,14 @@ class DownloadAttachmentProgressDialog : DialogFragment() {
     }
 
     private fun popBackStackWithError() {
-        val title = if (mainViewModel.isInternetAvailable.value == true) R.string.anErrorHasOccurred else R.string.noConnection
-        showSnackbar(title)
-        findNavController().popBackStack()
+        // We have to use the Flow instead of the LiveData here since the mainViewModel is not the same instance as the one in the
+        // MainActivity. So, instead of observing the LiveData in order to make it "active", we use the Flow.
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:83]: What LiveData are you talking about ? The one that you removed ?

File: app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt [hunk 10]
```diff
@@ -1091,5 +1102,6 @@ class MainViewModel @Inject constructor(
         private val DEFAULT_SELECTED_FOLDER = FolderRole.INBOX
         private const val REFRESH_DELAY = 2_000L // We add this delay because `etop` isn't always big enough.
         private const val MAX_REFRESH_DELAY = 6_000L
+        private const val TIMEOUT_MS_NETWORK_AVAILABILITY_MS = 500L
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/MainViewModel.kt:1105]: Unused.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt [hunk 11]
```diff
@@ -76,8 +79,14 @@ class DownloadAttachmentProgressDialog : DialogFragment() {
...
+        lifecycleScope.launch {
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:84]: Can you please specify the Dispatcher ?

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt (hunk 8 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt]: The `Dispatchers` is not important cause we're already in the main Dispatchers, but it's the wrong scope, use `viewLifecycleOwner.lifecycleScope` instead

File: app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt (hunk 6 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/ui/main/folder/ThreadListFragment.kt:532]: Hum...  I'm not a big fan, cause the launch return a job and i don't think that'll be used

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt (hunk 9 above)
Comment (tevincent) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:83]: The previous LiveData we had before that modification. But I can remove that explanation if that's misleading.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt (hunk 11 above)
Comment (sirambd) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt]: The `Dispatchers` is not important cause we're already in the main Dispatchers, but it's the wrong scope, use `viewLifecycleOwner.lifecycleScope` instead

**Review by sirambd — CHANGES_REQUESTED** [review]

**Review by sirambd — APPROVED** [review]

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt (hunk 9 above)
Comment (KevinBoulongne) [reply app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:83]: Yes you can.

File: app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt [hunk 12]
```diff
@@ -76,8 +79,11 @@ class DownloadAttachmentProgressDialog : DialogFragment() {
     }
 
     private fun popBackStackWithError() {
-        val title = if (mainViewModel.isInternetAvailable.value == true) R.string.anErrorHasOccurred else R.string.noConnection
-        showSnackbar(title)
-        findNavController().popBackStack()
+        viewLifecycleOwner.lifecycleScope.launch {
+            mainViewModel.isNetworkAvailable.first { it != null }?.let { isNetworkAvailable ->
+                showSnackbar(if (isNetworkAvailable) R.string.anErrorHasOccurred else R.string.noConnection)
```
Comment (KevinBoulongne) [inline app/src/main/java/com/infomaniak/mail/ui/main/thread/actions/DownloadAttachmentProgressDialog.kt:84]:
```suggestion
                showSnackbar(title = if (isNetworkAvailable) R.string.anErrorHasOccurred else R.string.noConnection)
```

**Review by KevinBoulongne — CHANGES_REQUESTED** [review]

**Review by KevinBoulongne — APPROVED** [review]

**Pull request merged**